libs/megaparse/src/api/megaparse_service.py

```diff
@@ -0,0 +1,190 @@
+            model = ChatAnthropic(
+                model_name=model_name,
+                api_key=os.getenv("ANTHROPIC_API_KEY"),  # type: ignore
+                timeout=60,
+                stop=None,
+            )
+        else:
+            raise ModelNotSupported
+        return model
+
+    async def parse_file(
+        self, file: FileInput, parser_config: ParseFileConfig
+    ) -> MPOutput:
+        logger.info(f"Parsing file {file.file_name} using config {parser_config}")
+        if not _check_free_memory():
+            return MPOutput(
+                output_type=MPOutputType.PARSE_ERR,
+                err=ParseError(
+                    mp_err_code=MPErrorType.MEMORY_LIMIT,
+                    message="high memory pressure on server",
+                ),
+                result=None,
+            )
+        try:
+            if parser_config.llm_model_name and parser_config.check_table:
+                model = self._init_model(parser_config.llm_model_name)
+
+            parser = self.parser_builder.build(parser_config)
+            megaparse = MegaParse(parser=parser)
+            _, extension = os.path.splitext(file.file_name)
+            file_stream = io.BytesIO(file.data)
+            result = await megaparse.aload(file=file_stream, file_extension=extension)
+            return MPOutput(output_type=MPOutputType.PARSE_OK, result=result)
+        except ParsingException as e:
+            return MPOutput(
+                output_type=MPOutputType.PARSE_ERR,
+                err=ParseError(
+                    mp_err_code=MPErrorType.PARSING_ERROR,
+                    message=str(e),
+                ),
+                result=None,
+            )
+
+        except Exception as e:
+            return MPOutput(
+                output_type=MPOutputType.PARSE_ERR,
+                err=ParseError(
+                    mp_err_code=MPErrorType.INTERNAL_SERVER_ERROR,
+                    message=str(e),
+                ),
+                result=None,
+            )
+
+    async def parse_url(
+        self,
+        url: str,
+    ):
+        assert self.browser
+        if url.endswith(".pdf"):
+            # TODO: don't have multiple clients'
+            async with httpx.AsyncClient() as client:
+                response = await client.get(url)
+            if response.status_code != 200:
+                raise Exception("Failed to download the file")
+
+            try:
+                megaparse = MegaParse(
+                    parser=UnstructuredParser(strategy=StrategyEnum.AUTO)
+                )
+                file_stream = io.BytesIO(response.content)
+                result = await megaparse.aload(
+                    file=file_stream, file_extension=FileExtension.PDF
+                )
+                return {"message": "File parsed successfully", "result": result}
+            except Exception as e:
+                raise Exception(str(e))
+        else:
+            page = await self.browser.new_page()
```

Comment:
Shouldn't we close the new page at some point, [here](https://playwright.dev/python/docs/pages) it says a new page is like a tab, so all the other pages are still somewhere in the memory (question)

Reply:
Nice catch !